File: src/main/java/Services.java
import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.Updates;
import org.bson.Document;
import org.bson.conversions.Bson;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.concurrent.locks.ReentrantLock;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Updates.inc;

public class Services {
    private static MongoCollection<Document> clientsCollection;
    private static final int maxWith=2000; //maximum daily withdrawal limit

    public static void setUpService(){
        clientsCollection= Database.getClients();
    }

    public static boolean validateCredentials(String username, String password){
        Document client = clientsCollection.find(new Document("username", username)).first();
        try{
            return client.get("password").equals(password);
        }
        catch (NullPointerException e){
            //client object is null pointer --> mongo returned zero results
            return  false;
        }
    }

    public static Double getAccountBalance(String username){
        return (Double) clientsCollection.find(new Document("username", username)).first().get("balance");
    }

    public static String getClientName(String username){

        Document client=clientsCollection.find(new Document("username", username)).first();

        Document logLogin = new Document().append("hour",
                LocalDateTime.now().getHour()).append("minute",
                LocalDateTime.now().getMinute()).append("second",
                LocalDateTime.now().getSecond());

        String subDocument=java.time.LocalDateTime.now().getDayOfMonth()+""+java.time.LocalDateTime.now().getMonthValue()+""+java.time.LocalDateTime.now().getYear();

        clientsCollection.updateOne(eq("username", client.get("username")), Updates.addToSet("loginsLog."+subDocument,logLogin));

        return client.get("firstName").toString()+" "+client.get("lastName");

    }

    public static boolean withdraw(String username, int amount){
        ReentrantLock userLock = Server.getUserLock(username);
        userLock.lock();

        Document client;

        try {
            client = clientsCollection.find(new Document("username", username)).first();
        }
        catch (NullPointerException e){
            //e.printStackTrace();
            return false;
        }

        if(amount>(double)client.get("balance") || amount>maxWith)
            return false;

        boolean transactionSuccess = false;

        try{

            String subDocument=java.time.LocalDateTime.now().getDayOfMonth()+""+java.time.LocalDateTime.now().getMonthValue()+""+java.time.LocalDateTime.now().getYear();

            ArrayList<Object> withdrawalsLog = (ArrayList<Object>) ((Document) client.get("withdrawalsLog")).get(subDocument);

            int sum=0;
            if (withdrawalsLog!=null) // -> mongo returns 0 results because the client has not made any withdrawals on the given date
                for(int i=0;i<withdrawalsLog.size();i++) {
                    Document transaction= (Document) withdrawalsLog.get(i);
                    sum+=(int) transaction.get("amount");
                }

            //System.out.println(sum); //For testing purposes
            if (amount+sum<=maxWith) {

                Bson incrementBalance = inc("balance", -amount);
                clientsCollection.updateOne(eq("username", client.get("username")), incrementBalance);

                Document logTransaction = new Document().append("hour",
                        java.time.LocalDateTime.now().getHour()).append("minute",
                        java.time.LocalDateTime.now().getMinute()).append("second",
                        java.time.LocalDateTime.now().getSecond()).append("amount",
                        amount);

                clientsCollection.updateOne(eq("username", client.get("username")), Updates.addToSet("withdrawalsLog."+subDocument,logTransaction));

                transactionSuccess=true;
            }

        }
        catch (Exception e){
            e.printStackTrace();
            return false;
        }
        finally {
            userLock.unlock();
        }
        return transactionSuccess;
    }

    public static boolean deposit(String username, int amount){
        try{

            Bson incrementBalance = inc("balance", amount);
            clientsCollection.updateOne(new Document("username", username),incrementBalance);

            String subDocument=java.time.LocalDateTime.now().getDayOfMonth()+""+java.time.LocalDateTime.now().getMonthValue()+""+java.time.LocalDateTime.now().getYear();

            Document logTransaction = new Document().append("hour",
                    java.time.LocalDateTime.now().getHour()).append("minute",
                    java.time.LocalDateTime.now().getMinute()).append("second",
                    java.time.LocalDateTime.now().getSecond()).append("amount",
                    amount);

            clientsCollection.updateOne(new Document("username", username), Updates.addToSet("depositsLog."+subDocument,logTransaction));

        }
        catch (NullPointerException e){
            return false;
        }
        return true;
    }

}

File: src/Services.py
import traceback
import datetime
import time
import threading

class Services():
    
    def __init__(self,clientsCollection):
        self.clientsCollection=clientsCollection
        self.maxWithdrawalAmount=2000
        self.userLocks = {aDocument["username"]:threading.Lock() for aDocument in self.clientsCollection.find()}
    
    def validateCredentials(self, username, password):
        clientDocument=self.clientsCollection.find_one({"username":username})
        
        if clientDocument and clientDocument["password"]==password:
            self.logTransaction(username,"loginsLog")
            return True
        return False

    def getName(self,username):
        clientDocument=self.clientsCollection.find_one({"username":username})
        return clientDocument["firstName"]+" "+clientDocument["lastName"] if clientDocument else ""

    def getAccountBalance(self,username):
        clientDocument=self.clientsCollection.find_one({"username":username})
        return str(clientDocument["balance"]) if clientDocument else ""

    def deposit(self,username,amount):
        try:
            self.clientsCollection.find_one_and_update({'username': username},{'$inc': {'balance': amount}})
            self.logTransaction(username,"depositsLog",amount)
        except:
            #traceback.print_exc()
            return (False,500)
        return (True,)
    
    def withdraw(self,username,amount):
        clientDocument=self.clientsCollection.find_one({"username":username})
        if not clientDocument:
            return (False,404)

        myLock=self.userLocks[clientDocument["username"]]

        myLock.acquire() #thread-safe
        try:
        
            if amount>self.maxWithdrawalAmount or amount>clientDocument["balance"]:
                return (False,400) #Bad request
            else:
                self.clientsCollection.find_one_and_update({'username': username},{'$inc': {'balance': -amount}})
                dateNow=datetime.datetime.now().date()
                logKey=f"{dateNow.day}{dateNow.month}{dateNow.year}"
                clientWithdrawalsToday=clientDocument["withdrawalsLog"].get(logKey)
                
                if not clientWithdrawalsToday:
                    self.logTransaction(username,"withdrawalsLog",amount)
                    return (True,)
                else:
                    total=sum(aSubDocument["amount"] for aSubDocument in clientWithdrawalsToday)
                    if total+amount>self.maxWithdrawalAmount:
                        return (False,400)
                    self.logTransaction(username,"withdrawalsLog",amount)
                    return (True,)
        except:
            #traceback.print_exc()
            return (False,500)
        finally:
            myLock.release()


    def logTransaction(self,username,logName,amount=None):
        timeNow=datetime.datetime.now()
        dateNow=timeNow.date()
        logKey=f"{dateNow.day}{dateNow.month}{dateNow.year}"
        if amount:
            self.clientsCollection.find_one_and_update({'username': username},{'$push': {f'{logName}.{logKey}': {"hour":timeNow.hour, "minute":timeNow.minute,"second":timeNow.second, "amount":amount}}})
        else:
            #logIns log -> no amount
            self.clientsCollection.find_one_and_update({'username': username},{'$push': {f'{logName}.{logKey}': {"hour":timeNow.hour, "minute":timeNow.minute,"second":timeNow.second}}})

if __name__ == '__main__':
    pass